Zend/zend_alloc.c

```diff
@@ -415,50 +415,77 @@ stderr_last_error(char *msg)
 /* OS Allocation */
 /*****************/
 
-#ifndef HAVE_MREMAP
-static void *zend_mm_mmap_fixed(void *addr, size_t size)
+static void zend_mm_munmap(void *addr, size_t size)
 {
 #ifdef _WIN32
-	return VirtualAlloc(addr, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+	MEMORY_BASIC_INFORMATION mbi;
+	if (VirtualQuery(addr, &mbi, sizeof(mbi)) == 0) {
+#if ZEND_MM_ERROR
+		stderr_last_error("VirtualQuery() failed");
+#endif
+	}
+	addr = mbi.AllocationBase;
+	
+	if (VirtualFree(addr, 0, MEM_RELEASE) == 0) {
+#if ZEND_MM_ERROR
+		stderr_last_error("VirtualFree() failed");
+#endif
+	}
 #else
-	int flags = MAP_PRIVATE | MAP_ANON;
-#if defined(MAP_EXCL)
-	flags |= MAP_FIXED | MAP_EXCL;
+	if (munmap(addr, size) != 0) {
+#if ZEND_MM_ERROR
+		fprintf(stderr, "\nmunmap() failed: [%d] %s\n", errno, strerror(errno));
 #endif
-	/* MAP_FIXED leads to discarding of the old mapping, so it can't be used. */
-	void *ptr = mmap(addr, size, PROT_READ | PROT_WRITE, flags /*| MAP_POPULATE | MAP_HUGETLB*/, ZEND_MM_FD, 0);
+	}
+#endif
+}
 
-	if (ptr == MAP_FAILED) {
-#if ZEND_MM_ERROR && !defined(MAP_EXCL)
-		fprintf(stderr, "\nmmap() failed: [%d] %s\n", errno, strerror(errno));
+#ifdef _WIN32
+static void *zend_mm_mmap_reserve(size_t size)
+{
+	void *ptr = VirtualAlloc(NULL, size, MEM_RESERVE, PAGE_READWRITE);
+	if (ptr == NULL) {
+#if ZEND_MM_ERROR
+		stderr_last_error("VirtualAlloc() reserve failed");
 #endif
 		return NULL;
-	} else if (ptr != addr) {
-		if (munmap(ptr, size) != 0) {
+	}
+	return ptr;
+}
+
+#ifndef _WIN64
+static void *zend_mm_mmap_reserve_try(void *addr, size_t size)
+{
+	void *ptr = VirtualAlloc(addr, size, MEM_RESERVE, PAGE_READWRITE);
+	if (ptr == NULL) {
+		/** ERROR_INVALID_ADDRESS is expected when addr - addr+size is not free */
+		if (GetLastError() != ERROR_INVALID_ADDRESS) {
 #if ZEND_MM_ERROR
-			fprintf(stderr, "\nmunmap() failed: [%d] %s\n", errno, strerror(errno));
+		stderr_last_error("VirtualAlloc() fixed reserve failed");
```

Comment:
```suggestion
			stderr_last_error("VirtualAlloc() fixed reserve failed");
```

Reply:
wdym?

Reply:
The statements within an `if` statement should have one more tab of indentation than the line with the `if`

Reply:
👍